### PR TITLE
add support for "-bypass-API" argument to "coq_makefile"

### PR DESCRIPTION
**Pull request by ghost**
From the commit message:
```
Plugin-writers can now use:
    
      -bypass-API
    
parameter with "coq_makefile".
    
The effect of that is that instead of
    
      -I API
    
the plugin will be compiled with:
    
      -I config" -I dev -I lib -I kernel -I library -I engine -I pretyping -I interp -I parsing -I proofs -I tactics -I toplevel -I printing -I intf -I grammar -I ide -I stm -I vernac
```

### CODE REVIEW

**Comment by matejkosik** [comment]
When reviewing, you can ignore these commits:
- `107d15e`  ... it is part of PR #700 
- `9264580` ... it is part of PR #652 

can be ignored (I guess).

File: lib/envars.mli [hunk 1]
```diff
@@ -76,7 +76,4 @@ val xdg_data_dirs   : (string -> unit) -> string list
 val xdg_dirs : warn : (string -> unit) -> string list
 
 (** {6 Prints the configuration information } *)
-val print_config : ?prefix_var_name:string -> out_channel -> unit
-
-(** Directories in which coq sources are found *)
-val coq_src_subdirs : string list
+val print_config : ?prefix_var_name:string -> out_channel -> string list -> unit
```
Comment (gares) [inline lib/envars.mli]: This is not OK in my humble opinion. In particular the print_config should not take a string list (with no label, no comments...). The config is Coq_config, potrentially overridden by env variables.  This third side-channel to change the config is a bit messy.

I see 2 options:
- pass to print_config the bypass_API boolean option, and have the function print the right set of dirs taken from Coq_config
- have print confing print all class of src_ directories, have coq_makefile put in the conf file if the API is bypassed, and finally have CoqMakefile.in do the if and use one or the other set of directories

File: toplevel/coqtop.ml [hunk 2]
```diff
@@ -621,7 +621,7 @@ let init_toplevel arglist =
       Spawned.init_channels ();
       Envars.set_coqlib ~fail:(fun msg -> CErrors.user_err Pp.(str msg));
       if !print_where then (print_endline(Envars.coqlib ()); exit(exitcode ()));
-      if !print_config then (Envars.print_config stdout; exit (exitcode ()));
+      if !print_config then (Envars.print_config stdout Coq_config.all_src_dirs; exit (exitcode ()));
```
Comment (gares) [inline toplevel/coqtop.ml]: I think this should also made to understand the bypass-API flag, or people writing their own Makefiles will have no choice.

**Comment by matejkosik** [comment]
@gares I was wondering how to customize the function. The option 1 you suggested makes more sense then what I managed to do so I went for that.

What needs to be done next?

**Comment by matejkosik** [comment]
@matejkosik : once #700 and #652 are merged, rebase this PR on trunk.

**Comment by gares** [comment]
Look good now.

It could be improved by having -bypass-API be _also_ understood by coqtop/coqc so that when -print-config is also passed one can obtain the same data that coq_makefile obtains (i.e. if one does not use coq_makefile can still get the appropriate set of directories using -print-config

**Comment by matejkosik** [comment]
@gares 
> It could be improved by having -bypass-API be also understood by coqtop/coqc so that when -print-config is also passed one can obtain the same data that coq_makefile obtains (i.e. if one does not use coq_makefile can still get the appropriate set of directories using -print-config

I haven't used `-print-config` option before. What does it do? Which command recognizes it?
```
mkosik@o1:~$ coqtop -print-config
Don't know what to do with -print-config
See -help for the list of supported options
mkosik@o1:~$ coqc -print-config
coqc: -print-config: no such file or directory
```

**Comment by gares** [comment]
typo on my side, the option is called `-config` and it is used by people writing their own Makefiles to get, among other things, the list of directories to use for `-I`.

**Comment by matejkosik** [comment]
@gares I've modified `coqc` and `coqtop` as you suggested:
```
mkosik@o1:~/git/coq/trunk1$ bin/coqtop -config
LOCAL=1
COQLIB=/home/mkosik/git/coq/trunk1//
DOCDIR=/home/mkosik/git/coq/trunk1/doc/
OCAMLFIND=/home/mkosik/.opam/4.02.3/bin/ocamlfind
CAMLP4=camlp5
CAMLP4O=/home/mkosik/.opam/4.02.3/bin/camlp5o
CAMLP4BIN=/home/mkosik/.opam/4.02.3/bin/
CAMLP4LIB=/home/mkosik/.opam/4.02.3/lib/camlp5
CAMLP4OPTIONS=-loc loc
HASNATDYNLINK=true
COQ_SRC_SUBDIRS=API lib plugins/btauto plugins/cc plugins/derive plugins/extraction plugins/firstorder plugins/fourier plugins/funind plugins/ltac plugins/micromega plugins/nsatz plugins/omega plugins/quote plugins/romega plugins/rtauto plugins/setoid_ring plugins/ssrmatching plugins/syntax plugins/xml

mkosik@o1:~/git/coq/trunk1$ bin/coqtop -bypass-API -config

LOCAL=1
COQLIB=/home/mkosik/git/coq/trunk1//
DOCDIR=/home/mkosik/git/coq/trunk1/doc/
OCAMLFIND=/home/mkosik/.opam/4.02.3/bin/ocamlfind
CAMLP4=camlp5
CAMLP4O=/home/mkosik/.opam/4.02.3/bin/camlp5o
CAMLP4BIN=/home/mkosik/.opam/4.02.3/bin/
CAMLP4LIB=/home/mkosik/.opam/4.02.3/lib/camlp5
CAMLP4OPTIONS=-loc loc
HASNATDYNLINK=true
COQ_SRC_SUBDIRS=config dev kernel library engine pretyping interp parsing proofs tactics toplevel printing intf grammar ide stm vernac API lib plugins/btauto plugins/cc plugins/derive plugins/extraction plugins/firstorder plugins/fourier plugins/funind plugins/ltac plugins/micromega plugins/nsatz plugins/omega plugins/quote plugins/romega plugins/rtauto plugins/setoid_ring plugins/ssrmatching plugins/syntax plugins/xml

mkosik@o1:~/git/coq/trunk1$ bin/coqc -config

LOCAL=1
COQLIB=/home/mkosik/git/coq/trunk1//
DOCDIR=/home/mkosik/git/coq/trunk1/doc/
OCAMLFIND=/home/mkosik/.opam/4.02.3/bin/ocamlfind
CAMLP4=camlp5
CAMLP4O=/home/mkosik/.opam/4.02.3/bin/camlp5o
CAMLP4BIN=/home/mkosik/.opam/4.02.3/bin/
CAMLP4LIB=/home/mkosik/.opam/4.02.3/lib/camlp5
CAMLP4OPTIONS=-loc loc
HASNATDYNLINK=true
COQ_SRC_SUBDIRS=API lib plugins/btauto plugins/cc plugins/derive plugins/extraction plugins/firstorder plugins/fourier plugins/funind plugins/ltac plugins/micromega plugins/nsatz plugins/omega plugins/quote plugins/romega plugins/rtauto plugins/setoid_ring plugins/ssrmatching plugins/syntax plugins/xml

mkosik@o1:~/git/coq/trunk1$ bin/coqc -bypass-API -config

LOCAL=1
COQLIB=/home/mkosik/git/coq/trunk1//
DOCDIR=/home/mkosik/git/coq/trunk1/doc/
OCAMLFIND=/home/mkosik/.opam/4.02.3/bin/ocamlfind
CAMLP4=camlp5
CAMLP4O=/home/mkosik/.opam/4.02.3/bin/camlp5o
CAMLP4BIN=/home/mkosik/.opam/4.02.3/bin/
CAMLP4LIB=/home/mkosik/.opam/4.02.3/lib/camlp5
CAMLP4OPTIONS=-loc loc
HASNATDYNLINK=true
COQ_SRC_SUBDIRS=config dev kernel library engine pretyping interp parsing proofs tactics toplevel printing intf grammar ide stm vernac API lib plugins/btauto plugins/cc plugins/derive plugins/extraction plugins/firstorder plugins/fourier plugins/funind plugins/ltac plugins/micromega plugins/nsatz plugins/omega plugins/quote plugins/romega plugins/rtauto plugins/setoid_ring plugins/ssrmatching plugins/syntax plugins/xml
```

**Comment by matejkosik** [comment]
**TODO** : rebase on `trunk` once #652 is merged.

**Comment by maximedenes** [comment]
@matejkosik Can you rebase?

**Comment by matejkosik** [comment]
Done.

**Comment by matejkosik** [comment]
Status update.

I see that our test-suite [fails](https://travis-ci.org/matejkosik/coq/builds/240865907) when we run it on MacOS.

@maximedenes Do we have an access to a machine where I can reproduce the problem in order to see exactly why it fails?

**Comment by maximedenes** [comment]
@matejkosik did you manage to fix the OSX issue today?

**Comment by matejkosik** [comment]
Sorry. Not yet. I am working on it.

**Comment by matejkosik** [comment]
@maximedenes Now it's [fixed](https://travis-ci.org/matejkosik/coq/builds/241447227).

**Comment by maximedenes** [comment]
@matejkosik I'll put your overlays in `ci-basic-overlay.sh` until the PRs are merged upstream.

**Comment by matejkosik** [comment]
That was I was hoping we can do.

Once merged, we can find a proper solution that will satisfy @JasonGross (for `bedrock` and `fiat`).

**Comment by maximedenes** [comment]
Merged, thanks!

**Pull request closed without merging**